### PR TITLE
Support immediately removing temp files

### PR DESCRIPTION
**Pull request by yonderblue**
A couple minor things too for your consideration @dim.

### CODE REVIEW

File: options.go [hunk 1]
```diff
@@ -21,6 +21,10 @@ type Options struct {
 	// By default os.TempDir() is used.
 	WorkDir string
 
+	// Keep temporary files until Close.
+	// Default: Immediately remove temporary files.
+	KeepFiles bool
```
Comment (dim) [inline options.go:26]: I don't think this option makes sense, I prefer not to add it and always remove immediately

File: tempfile.go [hunk 2]
```diff
@@ -4,33 +4,40 @@ import (
...
 
-func newTempWriter(dir string, compress Compression) (*tempWriter, error) {
-	f, err := ioutil.TempFile(dir, "extsort")
+func newTempWriter(dir string, compress Compression, keepFile bool) (*tempWriter, error) {
+	f, err := os.CreateTemp(dir, "extsort")
 	if err != nil {
 		return nil, err
 	}
+	if !keepFile {
+		// immediately remove for less chance of orphaning.
```
Comment (dim) [inline tempfile.go:27]: I don't think this is valid. Deleting open files may work on some file systems but it's not generally portable.

Comment (yonderblue) [reply tempfile.go:27]: True, I can add the build tags to make it only darwin plus linux.

File: options.go (hunk 1 above)
Comment (yonderblue) [reply options.go:26]: This was to keep your test that checks for compressed temporary files. Do you have a simple way to get around that?

File: tempfile.go (hunk 2 above)
Comment (yonderblue) [reply tempfile.go]: There is https://github.com/tmthrgd/tmpfile which would be a little better, but doesn't support the `pattern` at the moment, so what I have in here I think is simple enough.

File: tempfile_linux.go [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+func newTempFile(dir, pattern string, keepFile bool) (*os.File, error) {
+	f, err := os.CreateTemp(dir, pattern)
+	if err != nil {
+		return nil, err
+	}
+	if keepFile {
+		return f, nil
+	}
+	// immediately remove for less chance of orphaning.
+	if err := os.Remove(f.Name()); err != nil {
```
Comment (dim) [inline tempfile_linux.go:14]: Sorry to ask, but what's the purpose of this? All temp files will/should be removed once sorting is completed, so why would we remove them before they are written?

Comment (yonderblue) [reply tempfile_linux.go:14]: If a process crashes for whatever reason but the system doesn't restart for a long time (thus not clearing temp) you can build up a lot of orphan files, potentially wasting space or running out. Removing here as soon as it is created, means if the process goes away the file will be removed by the OS. The gap between creation and here is very small compared to the gap between creation and the end of sorting. This is the "old school" way for O_TMPFILE functionality before that was available.

**Comment by dim** [comment]
I have created #14 but the tests are still failing. I might be able to take a look early next week, unless you want to have a go

**Comment by yonderblue** [comment]
> I have created #14 but the tests are still failing. I might be able to take a look early next week, unless you want to have a go

I updated to just use the unix build tag which covers more. Since macOS has its own way to do the O_TMPFILE equivalent this is needed, even if it did use O_TMPFILE just for linux.
I'd recommend just keeping this as it is simpler and does the job.

**Comment by dim** [comment]
> > I have created #14 but the tests are still failing. I might be able to take a look early next week, unless you want to have a go
> 
> I updated to just use the unix build tag which covers more. Since macOS has its own way to do the O_TMPFILE equivalent this is needed, even if it did use O_TMPFILE just for linux. I'd recommend just keeping this as it is simpler and does the job.

I still feel that `"immediately remove for less chance of orphaning"` is a bit of a hack. I will take a proper look on Monday!

**Comment by dim** [comment]
Sorry, forgot about that one, do you mind rebasing?

**Comment by yonderblue** [comment]
> Sorry, forgot about that one, do you mind rebasing?

Yup no problem, rebased 👍

**Pull request merged**